subsys/net/ip/net_pkt.c

```diff
@@ -1367,10 +1381,15 @@ pkt_alloc_with_buffer(struct k_mem_slab *slab,
 
 	net_pkt_set_family(pkt, family);
 
-	if (timeout != K_NO_WAIT && timeout != K_FOREVER) {
-		u32_t diff = k_uptime_get_32() - alloc_start;
+	if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT) &&
+	    !K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+		s64_t remaining = end - z_tick_get();
 
-		timeout -= MIN(timeout, diff);
+		if (remaining <= 0) {
+			timeout = K_NO_WAIT;
+		} else {
+			timeout = Z_TIMEOUT_TICKS(remaining);
```

Comment:
FWIW: "K_NO_WAIT" is just a timeout of zero, and the Z_TIMEOUT_*() APIs include the clamp for you.  You can just use Z_TIMEOUT_TICKS(end - z_tick_get()) if you like.